rfd/0058-desktop-file-transfer.md

```diff
@@ -0,0 +1,143 @@
+---
+authors: Isaiah Becker-Mayer (isaiah@goteleport.com)
+state: draft
+---
+
+# RFD 58 - Desktop Access - File Transfer
+
+File transfer for Teleport Desktop Access is a feature whose implementation includes navigating the constraints imposed by the RDP protocol, the limitations of the browser, Teleport's architecture, and storage and bandwidth implications. The purpose of this document is to lay out the approaches available to us given these constraints, and compare them in terms of user experience, difficulty of technical implementation, and other relevant criteria in order determine which approach we should take. Therefore, some of the finer technical details of each approach will be left in the abstract in favor of keeping this document concise and related to its core purpose.
+
+## RDP (Server Side) Possibilities and Data Flow
+
+RDP supports two means of transferring files between systems:
+
+1. The clipboard virtual channel extension ([[MS-RDPECLIP]](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip)).
+2. The file system virtual channel extension ([[MS-RDPEFS]](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs)).
+
+### Clipboard Virtual Channel Extension (RDP Option 1)
+
+The data flow when using the clipboard virtual channel extension (henceforth referred to as RDP Option 1) includes the one described in [Data Flow and Delayed Rendering](https://github.com/gravitational/teleport/blob/master/rfd/0049-desktop-clipboard.md#data-flow-and-delayed-rendering) in RFD 0049, but with extra steps appended. In the case that a file, directory, or multiple files/directories have been cut or copied onto the shared clipboard, the `Format Data Response PDU` doesn't respond with the file data itself, but instead sends a [`Packed File List`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/3570c2e4-cdd7-4460-8a7e-1a4595f5ebdc) which contains a list of filenames and associated metadata of the files on the clipboard. The files themselves aren't transferred until the shared clipboard owner receives a [`File Contents Request PDU`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/cbc851d3-4e68-45f4-9292-26872a9209f2), at which point it sends back the file data in the form of [`File Contents Response PDU`(`s`)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/df87c178-ab02-471a-acde-bb921aa1af85).
+
+File transfer's are limited to files less than 4GB unless huge file support is enabled (`CB_HUGE_FILE_SUPPORT_ENABLED`).
+
+### File System Virtual Channel Extension (RDP Option 2)
+
+From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all functionality on that directory is instigated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.
```

Comment:
```suggestion
From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all actions on that directory are initiated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.
```

Reworded this a bit, but it's still not quite clear to me what you're describing here.